quickwit/quickwit-ingest/src/ingest_v2/mod.rs

```diff
@@ -139,6 +141,41 @@ impl DocBatchV2Builder {
         };
         Some(doc_batch)
     }
+
+    pub fn json_writer(self) -> JsonDocBatchV2Builder {
+        JsonDocBatchV2Builder {
+            doc_uids: self.doc_uids,
+            doc_buffer: self.doc_buffer.writer(),
+            doc_lengths: self.doc_lengths,
+        }
+    }
+}
+
+/// Wrapper around the batch builder that can add Serialize structs without an extra copy
+pub struct JsonDocBatchV2Builder {
+    doc_uids: Vec<DocUid>,
+    doc_buffer: Writer<BytesMut>,
+    doc_lengths: Vec<u32>,
+}
+
+impl JsonDocBatchV2Builder {
+    pub fn add_doc(&mut self, doc_uid: DocUid, payload: impl Serialize) -> serde_json::Result<()> {
+        let old_len = self.doc_buffer.get_ref().len();
+        serde_json::to_writer(&mut self.doc_buffer, &payload)?;
+        let new_len = self.doc_buffer.get_ref().len();
+        let written_len = new_len - old_len;
+        self.doc_uids.push(doc_uid);
+        self.doc_lengths.push(written_len as u32);
+        Ok(())
+    }
+
+    pub fn into_inner(self) -> DocBatchV2Builder {
```

Comment:
I would not call that `into_inner`. We could use `DocBatchV2Builder::from` (and implement the trait)